discovery-provider/src/tasks/entity_manager/utils.py

```diff
@@ -308,12 +308,19 @@ def parse_metadata(metadata, action, entity_type):
 
         cid = data["cid"]
         metadata_json = data["data"]
-        _, metadata_format = get_metadata_type_and_format(entity_type)
-        formatted_json = get_metadata_from_json(metadata_format, metadata_json)
 
-        # Only index valid changes
-        if formatted_json == metadata_format:
-            raise IndexingValidationError("no valid metadata changes detected")
+        # Don't format metadata for UPDATEs
+        # This is to support partial updates 
+        # Individual entities are responsible for updating existing records with metadata
+        if action != Action.UPDATE:
+            _, metadata_format = get_metadata_type_and_format(entity_type)
+            formatted_json = get_metadata_from_json(metadata_format, metadata_json)
+
+            # Only index valid changes
+            if formatted_json == metadata_format:
+                raise IndexingValidationError("no valid metadata changes detected")
+        else:
+            formatted_json = metadata_json
```

Comment:
I love the idea, but I assume the cid will now be a hash of the partial metadata and is subsequently saved as the `metadata_multihash` on the user / track / playlist. the client still [queries this cid](https://github.com/AudiusProject/audius-client/blob/83feb573d4602135c870399e000641426ceb6485/packages/common/src/services/audius-backend/AudiusBackend.ts#L421) from discovery to retrieve some things that aren't indexed by discovery -- mainly collectibles and wallet info I believe. i also think the client caches all values (like user description, etc) from the blob. if we have partial updates, we'll be missing necessary info in the blob referenced by the cid.

we currently save the cid, cid type (user, track, or playlist_data), and corresponding metadata blob in the `cid_data` table in discovery. while iterating through every tx in a block, if the entity type is USER, TRACK, or PLAYLIST, we [write to a cid_metadata dict ](https://github.com/AudiusProject/audius-protocol/blob/07bd08f93e2b43626aca2adaf8fa2aa2c9c43c97/discovery-provider/src/tasks/entity_manager/entity_manager.py#L167)mapping the cid -> the corr. blob.

even with partial updates, we should support having every current metadata value + the updated values in the corresponding metadata blob for when the client queries the cid. not ideal, but unfortunately necessary until we support indexing every field in the metadata blob + backfill this info.

one way we could make partial updates work with this constraint: move the `cid_metadata` and `cid_type` inserts to each event handler (e.g. pass as a param to `update_user`, `update_track`, etc.) and set `cid_metadata[params.metadata_cid]` to the entire current record in the db with the updates applied. the cid wouldn't be the hash of the right blob anymore but i dont think that matters.

Reply:
Amazing feedback, thank you!

Makes total sense, ideally we would move away from querying the cid metadata blob. But for now I'll try to write the db record + update as the metadata blob. Is there still a need to generate the cid client side? If not I could generate the cid from the updated db record, so that it's a hash of the correct data. If we only hash the partial update metadata, there could be a situation where we fetch the wrong blob I think

Reply:
oh yea great point ^ and i think there are a few places we have to be careful about partial updates because not everything can be partial atm such as playlist ordering. i wonder if we can just do a merge upstream and then hash it in the indexer instead.

Reply:
yea @sliptype i wrote the comment at the same time and was thinking the same haha. lmk if i can help w this

Reply:
missed this thread, sorry for the late reply - i think hashing server-side is totally fine as long as the client knows the cid after the update

Reply:
No worries! Cool yeah the client just gets the cid from the user endpoint as `metadata_multihash` so I think we should be good there